File: src/class037/C3_PathSumII.cpp
#include <iostream>
#include <vector>
using namespace std;

// 定义树的节点结构
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

// 收集累加和等于aim的所有路径
// 测试链接 : https://leetcode.cn/problems/path-sum-ii/
class Code03_PathSumII {
public:
    // 提交以下的方法
    vector<vector<int>> pathSum(TreeNode* root, int aim) {
        vector<vector<int>> ans;
        if (root != nullptr) {
            vector<int> path;
            f(root, aim, 0, path, ans);
        }
        return ans;
    }

private:
    void f(TreeNode* cur, int aim, int sum, vector<int>& path, vector<vector<int>>& ans) {
        if (cur->left == nullptr && cur->right == nullptr) {
            // 叶节点
            if (cur->val + sum == aim) {
                path.push_back(cur->val); // 加入自己
                copy(path, ans); // path生成好
                path.pop_back(); // 从path中移除最后一个
            }
        } else {
            // 不是叶节点
            path.push_back(cur->val);
            if (cur->left != nullptr) { // 左非空
                f(cur->left, aim, sum + cur->val, path, ans);
            }
            if (cur->right != nullptr) { // 右非空
                f(cur->right, aim, sum + cur->val, path, ans);
            }
            path.pop_back(); //删掉自己，恢复现场
        }
    }

    void copy(const vector<int>& path, vector<vector<int>>& ans) {
        ans.push_back(vector<int>(path.begin(), path.end()));
    }
};

File: src/class037/C3_PathSumII.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Code03_PathSumII:
    def pathSum(self, root, aim):
        """
        收集累加和等于aim的所有路径
        测试链接: https://leetcode.cn/problems/path-sum-ii/
        """
        ans = []
        if root is not None:
            path = []
            self.f(root, aim, 0, path, ans)
        return ans

    def f(self, cur, aim, sum, path, ans):
        if cur.left is None and cur.right is None:
            # 叶节点
            if cur.val + sum == aim:
                path.append(cur.val)
                ans.append(path.copy())
                path.pop()
        else:
            # 不是叶节点
            path.append(cur.val)
            if cur.left is not None:
                self.f(cur.left, aim, sum + cur.val, path, ans)
            if cur.right is not None:
                self.f(cur.right, aim, sum + cur.val, path, ans)
            path.pop()

# 示例使用
if __name__ == "__main__":
    # 构造一个简单的测试树：      5
    #                         /     \
    #                        4       8
    #                       /       / \
    #                      11      13  4
    #                     /  \          \
    #                    7    2          1
    root = TreeNode(5)
    root.left = TreeNode(4)
    root.right = TreeNode(8)
    root.left.left = TreeNode(11)
    root.left.left.left = TreeNode(7)
    root.left.left.right = TreeNode(2)
    root.right.left = TreeNode(13)
    root.right.right = TreeNode(4)
    root.right.right.right = TreeNode(1)

    # 创建实例并调用函数
    solver = Code03_PathSumII()
    result = solver.pathSum(root, 22)
    print("路径总和为22的所有路径:", result)
